managers/bosh-manager/DirectorService.js

```diff
@@ -0,0 +1,1116 @@
+    logger.info(`Checking rate limits against bosh for deployment `);
+    const previousValues = _.get(params, 'previous_values');
+    const action = _.isPlainObject(previousValues) ? CONST.OPERATION_TYPE.UPDATE : CONST.OPERATION_TYPE.CREATE;
+    const scheduled = _.get(params, 'scheduled') || false;
+    const runImmediately = _.get(params, 'parameters._runImmediately') || false;
+    _.omit(params, 'scheduled');
+    _.omit(params, 'parameters._runImmediately');
+
+    if (!config.enable_bosh_rate_limit || runImmediately) {
+      return this._createOrUpdateDeployment(deploymentName, params, args, scheduled)
+        .then(taskId => {
+          return {
+            cached: false,
+            task_id: taskId
+          };
+        });
+    }
+    const decisionMaker = {
+      'shouldRunNow': false,
+      'cached': false
+    };
+    return this.executePolicy(scheduled, action, deploymentName)
+      .then(checkResults => this.enqueueOrTrigger(checkResults.shouldRunNow, scheduled, deploymentName))
+      .tap(res => {
+        decisionMaker.shouldRunNow = res.shouldRunNow;
+        decisionMaker.cached = res.cached;
+      })
+      .then(res => {
+        if (res.enqueue) {
+          // stagger here by putting it into etcd cache and return promise
+          return boshOperationQueue.saveDeployment(this.plan.id, deploymentName, params, args)
+            .then(() => {
+              decisionMaker.cached = true;
+              throw new DeploymentDelayed(deploymentName);
+            });
+        }
+      })
+      .then(() => {
+        if (decisionMaker.shouldRunNow && decisionMaker.cached) {
+          //the deployment was cached in etcd earlier; remove it from cache and proceed
+          return boshOperationQueue.deleteDeploymentFromCache(deploymentName);
+        }
+      })
+      .then(() => this._createOrUpdateDeployment(deploymentName, params, args, scheduled))
+      .then(taskId => {
+        const out = {
+          'cached': false
+        };
+        if (decisionMaker.cached) {
+          return boshOperationQueue.saveBoshTask(this.getInstanceGuid(deploymentName), taskId)
+            .then(() => {
+              out.cached = true;
+              out.task_id = taskId;
+              return out;
+            });
+        } else {
+          out.task_id = taskId;
+          return out;
+        }
+      })
+      .catch(DeploymentDelayed, e => {
+        logger.info(`Deployment ${deploymentName} delayed- this should be picked up later for processing`, e);
+        return {
+          'cached': true
+        };
+      })
+      .catch(err => {
+        logger.error(`Error in deployment for ${deploymentName}`, err);
+        throw err;
+      });
+  }
+
+  _createOrUpdateDeployment(deploymentName, params, args, scheduled) {
+    const previousValues = _.get(params, 'previous_values');
+    const action = _.isPlainObject(previousValues) ? CONST.OPERATION_TYPE.UPDATE : CONST.OPERATION_TYPE.CREATE;
```

Comment:
Might be not a good comment for this PR. But previous_values is not a mandatory parameter. Better to pass a parameter in function determine create or update.

Reply:
`isPlainObject ` method anyway returns false for undefined values so it is handled. Didn't understand what you mean.